File: lambda.cpp
#include "lambda.h"
#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <map>
#ifdef _WIN32
#include <Windows.h>
#endif

const std::string symbols = "xyzabcdefg";
const std::string lambda = "λ";
std::string pixels[16][16] = {
    {" ", "𜺨", "𜺫", "🮂", "𜴀", "▘", "𜴁", "𜴂", "𜴃", "𜴄", "▝", "𜴅", "𜴆", "𜴇", "𜴈", "▀"},
    {"𜴉", "𜴊", "𜴋", "𜴌", "🯦", "𜴍", "𜴎", "𜴏", "𜴐", "𜴑", "𜴒", "𜴓", "𜴔", "𜴕", "𜴖", "𜴗"},
    {"𜴘", "𜴙", "𜴚", "𜴛", "𜴜", "𜴝", "𜴞", "𜴟", "🯧", "𜴠", "𜴡", "𜴢", "𜴣", "𜴤", "𜴥", "𜴦"},
    {"𜴧", "𜴨", "𜴩", "𜴪", "𜴫", "𜴬", "𜴭", "𜴮", "𜴯", "𜴰", "𜴱", "𜴲", "𜴳", "𜴴", "𜴵", "🮅"},
    {"𜺣", "𜴶", "𜴷", "𜴸", "𜴹", "𜴺", "𜴻", "𜴼", "𜴽", "𜴾", "𜴿", "𜵀", "𜵁", "𜵂", "𜵃", "𜵄"},
    {"▖", "𜵅", "𜵆", "𜵇", "𜵈", "▌", "𜵉", "𜵊", "𜵋", "𜵌", "▞", "𜵍", "𜵎", "𜵏", "𜵐", "▛"},
    {"𜵑", "𜵒", "𜵓", "𜵔", "𜵕", "𜵖", "𜵗", "𜵘", "𜵙", "𜵚", "𜵛", "𜵜", "𜵝", "𜵞", "𜵟", "𜵠"},
    {"𜵡", "𜵢", "𜵣", "𜵤", "𜵥", "𜵦", "𜵧", "𜵨", "𜵩", "𜵪", "𜵫", "𜵬", "𜵭", "𜵮", "𜵯", "𜵰"},
    {"𜺠", "𜵱", "𜵲", "𜵳", "𜵴", "𜵵", "𜵶", "𜵷", "𜵸", "𜵹", "𜵺", "𜵻", "𜵼", "𜵽", "𜵾", "𜵿"},
    {"𜶀", "𜶁", "𜶂", "𜶃", "𜶄", "𜶅", "𜶆", "𜶇", "𜶈", "𜶉", "𜶊", "𜶋", "𜶌", "𜶍", "𜶎", "𜶏"},
    {"▗", "𜶐", "𜶑", "𜶒", "𜶓", "▚", "𜶔", "𜶕", "𜶖", "𜶗", "▐", "𜶘", "𜶙", "𜶚", "𜶛", "▜"},
    {"𜶜", "𜶝", "𜶞", "𜶟", "𜶠", "𜶡", "𜶢", "𜶣", "𜶤", "𜶥", "𜶦", "𜶧", "𜶨", "𜶩", "𜶪", "𜶫"},
    {"▂", "𜶬", "𜶭", "𜶮", "𜶯", "𜶰", "𜶱", "𜶲", "𜶳", "𜶴", "𜶵", "𜶶", "𜶷", "𜶸", "𜶹", "𜶺"},
    {"𜶻", "𜶼", "𜶽", "𜶾", "𜶿", "𜷀", "𜷁", "𜷂", "𜷃", "𜷄", "𜷅", "𜷆", "𜷇", "𜷈", "𜷉", "𜷊"},
    {"𜷋", "𜷌", "𜷍", "𜷎", "𜷏", "𜷐", "𜷑", "𜷒", "𜷓", "𜷔", "𜷕", "𜷖", "𜷗", "𜷘", "𜷙", "𜷚"},
    {"▄", "𜷛", "𜷜", "𜷝", "𜷞", "▙", "𜷟", "𜷠", "𜷡", "𜷢", "▟", "𜷣", "▆", "𜷤", "𜷥", "█"},
};

Expression::~Expression() {}

std::deque<char> Abstraction::symbol_stack;

Abstraction::Abstraction(std::unique_ptr<Expression> e): expression(std::move(e)) {}

std::unique_ptr<Expression> Abstraction::reduce(bool* changed) {
    std::unique_ptr<Expression> body = this->expression->reduce(changed);
    if (body != nullptr) {
        this->expression = std::move(body);
    }
    return nullptr;
}

std::unique_ptr<Expression> Abstraction::copy() const {
    std::unique_ptr<Expression> abstraction(new Abstraction(this->expression->copy()));
    return abstraction;
}

std::string Abstraction::tostring() {
    Abstraction::symbol_stack.emplace_front(symbols[Abstraction::symbol_stack.size()]);
    std::string body = this->expression->tostring();
    if (body[0] == '(') {
        body = body.substr(1, body.length() - 2);
    }
    std::ostringstream stream;
    stream << "(" << lambda << Abstraction::symbol_stack[0] << '.' << body << ')';
    Abstraction::symbol_stack.pop_front();
    return stream.str();
}

std::unique_ptr<Expression> Abstraction::beta(int index, const Expression& value) {
    std::unique_ptr<Expression> body = this->expression->beta(index + 1, value);
    if (body != nullptr) {
        this->expression = std::move(body);
    }
    return nullptr;
}

void Abstraction::shift_free(int index, int shift) {
    this->expression->shift_free(index + 1, shift);
}

Application::Application(
    std::unique_ptr<Expression> l,
    std::unique_ptr<Expression> r): left(std::move(l)), right(std::move(r)) {}

std::unique_ptr<Expression> Application::reduce(bool* changed) {
    Abstraction *abstraction = dynamic_cast<Abstraction*>(this->left.get());
    if (abstraction != nullptr) {
        std::unique_ptr<Expression> body = std::move(abstraction->expression);
        this->right->shift_free(0, 1);
        *changed = true;

        std::unique_ptr<Expression> result = body->beta(0, *this->right);
        if (result != nullptr) {
            result->shift_free(0, -1);
            return result;
        }
        body->shift_free(0, -1);
        return body;
    }

    bool left_changed;
    std::unique_ptr<Expression> left = this->left->reduce(&left_changed);
    if (left_changed) {
        if (left != nullptr) {
            this->left = std::move(left);
        }
        *changed = true;
        return nullptr;
    }

    bool right_changed;
    std::unique_ptr<Expression> right = this->right->reduce(&right_changed);
    if (right_changed) {
        if (right != nullptr) {
            this->right = std::move(right);
        }
        *changed = true;
        return nullptr;
    }

    *changed = false;
    return nullptr;
}

std::unique_ptr<Expression> Application::copy() const {
    std::unique_ptr<Expression> application(new Application(this->left->copy(), this->right->copy()));
    return application;
}

std::string Application::tostring() {
    std::string left = this->left->tostring();
    std::string right = this->right->tostring();
    std::ostringstream stream;
    if (left.length() == 1 && right.length() == 1) {
        stream << '(' << left << ' ' << right << ')';
        return stream.str();
    } else if (left[0] == '(' && left.substr(1, lambda.size()) != lambda) {
        if (right.length() == 1) {
            stream << left.substr(0, left.length() - 1) << ' ' << right << ')';
        } else {
            stream << '(' << left << right << ')';
        }
        return stream.str();
    }
    stream << '(' << left << right << ')';
    return stream.str();
}

std::unique_ptr<Expression> Application::beta(int index, const Expression& value) {
    std::unique_ptr<Expression> left = this->left->beta(index, value);
    if (left != nullptr) {
        this->left = std::move(left);
    }
    std::unique_ptr<Expression> right = this->right->beta(index, value);
    if (right != nullptr) {
        this->right = std::move(right);
    }
    return nullptr;
}

void Application::shift_free(int index, int shift) {
    this->left->shift_free(index, shift);
    this->right->shift_free(index, shift);
}

Variable::Variable(int i): index(i) {}

std::unique_ptr<Expression> Variable::reduce(bool* changed) {
    *changed = false;
    return nullptr;
}

std::unique_ptr<Expression> Variable::copy() const {
    std::unique_ptr<Expression> variable(new Variable(this->index));
    return variable;
}

std::string Variable::tostring() {
    std::string string(1, Abstraction::symbol_stack[this->index]);
    return string;
}

std::unique_ptr<Expression> Variable::beta(int index, const Expression& value) {
    if (this->index == index) {
        std::unique_ptr<Expression> copy = value.copy();
        copy->shift_free(0, index);
        return copy;
    }
    return nullptr;
}

void Variable::shift_free(int index, int shift) {
    if (this->index >= index) {
        this->index += shift;
    }
}

std::unique_ptr<Expression> parse_string(std::string expr) {
    std::string string = expr;
    size_t index = 0;
    while (true) {
        index = string.find(lambda, index);
        if (index == std::string::npos) {
            break;
        }
        string.replace(index, lambda.size(), "#");
        index += 1;
    }

    if (string[0] == '#') {
        Abstraction::symbol_stack.emplace_front(string[1]);
        std::unique_ptr<Expression> abstraction(new Abstraction(parse_string(string.substr(3))));
        Abstraction::symbol_stack.pop_front();
        return abstraction;
    } else if (string.length() == 1) {
        int index = -1;
        for (int i = 0; i < Abstraction::symbol_stack.size(); i++) {
            if (Abstraction::symbol_stack[i] == string[0]) {
                index = i;
                break;
            }
        }

        std::unique_ptr<Expression> variable(new Variable(index));
        return variable;
    }

    std::deque<std::unique_ptr<Expression>> applications;
    while (string.length() > 0) {
        if (string[0] == '(') {
            int end = 0;
            int brackets = 0;
            for (int i = 0; i < string.length(); i++) {
                if (string[i] == '(') {
                    brackets++;
                } else if (string[i] == ')') {
                    brackets--;
                }
                if (brackets == 0) {
                    end = i;
                    break;
                }
            }
            applications.emplace_back(parse_string(string.substr(1, end - 1)));
            string = string.substr(end + 1);
        } else {
            applications.emplace_back(parse_string(string.substr(0, 1)));
            string = string.substr(1);
            if (string[0] == ' ') {
                string = string.substr(1);
            }
        }
    }

    while (applications.size() > 1) {
        std::unique_ptr<Expression> application(
            new Application(std::move(applications[0]), std::move(applications[1])));
        applications.pop_front();
        applications[0] = std::move(application);
    }
    return std::move(applications[0]);
}

void dump(std::unique_ptr<Expression>& expression, bool alternative = true) {
    std::vector<int> variables;
    std::vector<int> application_depths;
    std::map<int, std::vector<std::pair<int, int>>> abstractions;
    std::vector<int> abstraction_stack;
    std::vector<std::tuple<int, int, int>> applications;
    std::vector<std::pair<int, int>> branching_points;

    std::vector<std::pair<const Expression*, int>> stack;
    stack.emplace_back(expression.get(), 0);
    while (stack.size() > 0) {
        std::pair<const Expression*, int> pair = stack.back();
        stack.pop_back();
        if (pair.second == 1) {
            int start = abstraction_stack.back();
            abstraction_stack.pop_back();
            int end = variables.size() - 1;
            int index = abstraction_stack.size();
            if (abstractions.find(index) == abstractions.end()) {
                abstractions[index] = {};
            }
            abstractions[index].emplace_back(start, end);
        } else if (pair.second == 2) {
            std::pair<int, int> right = branching_points.back();
            branching_points.pop_back();
            int b = right.first;
            std::pair<int, int> left = branching_points.back();
            branching_points.pop_back();
            int a = (alternative) ? left.second : left.first;

            int depth = std::max(application_depths[a], application_depths[b]);
            applications.emplace_back(a, b, depth);
            application_depths[a] = depth + 1;
            application_depths[b] = depth + 1;
            branching_points.emplace_back(a, b);
        } else {
            const Abstraction *abstraction = dynamic_cast<const Abstraction*>(pair.first);
            if (abstraction != nullptr) {
                abstraction_stack.emplace_back(variables.size());
                stack.emplace_back(nullptr, 1);
                stack.emplace_back(abstraction->expression.get(), 0);
                continue;
            }

            const Application *application = dynamic_cast<const Application*>(pair.first);
            if (application != nullptr) {
                stack.emplace_back(nullptr, 2);
                stack.emplace_back(application->right.get(), 0);
                stack.emplace_back(application->left.get(), 0);
                continue;
            }

            const Variable *variable = dynamic_cast<const Variable*>(pair.first);
            if (variable != nullptr) {
                variables.emplace_back(abstraction_stack.size() - variable->index - 1);
                application_depths.emplace_back(abstraction_stack.size());
                branching_points.emplace_back(variables.size() - 1, variables.size() - 1);
                continue;
            }
        }
    }

    if (!alternative) {
        application_depths[0]++;
    }

    int width = variables.size() * 4;
    int height = *std::max_element(application_depths.begin(), application_depths.end()) * 2;

    int **canvas = new int*[height];
    for (int row = 0; row < height; row++) {
        canvas[row] = new int[width];
        for (int col = 0; col < width; col++) {
            canvas[row][col] = 0;
        }
    }

    for (int x = 0; x < variables.size(); x++) {
        for (int y = variables[x] * 2; y < application_depths[x] * 2 - 1; y++) {
            canvas[y][x * 4 + 1] = 1;
        }
    }

    for (auto it = abstractions.begin(); it != abstractions.end(); it++) {
        for (int i = 0; i < it->second.size(); i++) {
            std::pair<int, int> abstraction = it->second[i];
            int depth = it->first;
            int start = abstraction.first;
            int end = abstraction.second;
            for (int x = start * 4; x < end * 4 + 3; x++) {
                canvas[depth * 2][x] = 1;
            }
        }
    }

    for (int i = 0; i < applications.size(); i++) {
        int start = std::get<0>(applications[i]);
        int end = std::get<1>(applications[i]);
        int depth = std::get<2>(applications[i]);
        for (int x = start * 4 + 1; x < end * 4 + 2; x++) {
            canvas[depth * 2][x] = 1;
        }
    }

    // std::string pixels[2][2] = {{" ", "▄"}, {"▀", "█"}};
    // std::string graph = "";
    // for (int y = 0; y < height; y += 2) {
    //     for (int x = 0; x < width; x++) {
    //         graph.append(pixels[canvas[y][x]][canvas[y + 1][x]]);
    //     }
    //     graph.append("\n");
    // }
    // std::cout << graph << std::flush;

    std::string graph = "";
    for (int y = 0; y < height; y += 4) {
        for (int x = 0; x < width; x += 2) {
            int top = canvas[y][x] + canvas[y][x+1]*2 + canvas[y+1][x]*4 + canvas[y+1][x+1]*8;
            int bottom;
            if (y == height - 2) {
                bottom = 0;
            } else {
                bottom = canvas[y+2][x] + canvas[y+2][x+1]*2 + canvas[y+3][x]*4 + canvas[y+3][x+1]*8;
            }
            graph.append(pixels[bottom][top]);
        }
        graph.append("\n");
    }
    std::cout << graph << std::flush;

    for (int y = 0; y < height; y++) {
        delete[] canvas[y];
    }
    delete[] canvas;
}

#define L(expr) std::unique_ptr<Expression>(new Abstraction(expr))
#define A(l, r) std::unique_ptr<Expression>(new Application(l, r))
#define V(idx) std::unique_ptr<Expression>(new Variable(idx))
#define C(expr) expr->copy()

int main(int argc, char *argv[]) {
#ifdef _WIN32
    SetConsoleOutputCP(CP_UTF8);
#endif
    auto s = L(L(L(A(A(V(2), V(0)), A(V(1), V(0))))));
    auto k = L(L(V(1)));
    auto i = L(V(0));
    auto iota1 = A(A(C(s), A(A(C(s), C(i)), A(C(k), C(s)))), A(C(k), C(k)));

    auto s2 = A(C(iota1), A(C(iota1), A(C(iota1), A(C(iota1), C(iota1)))));
    auto k2 = A(C(iota1), A(C(iota1), A(C(iota1), C(iota1))));
    auto i2 = A(C(iota1), C(iota1));
    auto iota2 = A(A(C(s2), A(A(C(s2), C(i2)), A(C(k2), C(s2)))), A(C(k2), C(k2)));

    auto expression = iota2->copy();
    std::cout << expression->tostring() << std::endl;

    bool changed = true;
    while (changed) {
        dump(expression);
        std::unique_ptr<Expression> new_expression = expression->reduce(&changed);
        if (new_expression != nullptr) {
            expression = std::move(new_expression);
        }
    }
    return 0;
}

File: lambda.py
from abc import ABCMeta, abstractmethod

symbols = "xyzabcdef"

class Expression(metaclass=ABCMeta):
    @abstractmethod
    def reduce(self): pass

    @abstractmethod
    def copy(self): pass

    @abstractmethod
    def __repr__(self): pass

    @abstractmethod
    def beta(self, index, value): pass

    @abstractmethod
    def shift_free(self, index, shift): pass

class Abstraction(Expression):
    symbol_stack = []

    def __init__(self, expression):
        self.expression = expression

    def reduce(self):
        expression = self.expression.reduce()
        if expression is not None:
            return Abstraction(expression)
        return None

    def copy(self):
        return Abstraction(self.expression.copy())

    def __repr__(self):
        Abstraction.symbol_stack.insert(0, symbols[len(Abstraction.symbol_stack)])
        body = repr(self.expression)
        if isinstance(self.expression, (Abstraction, Application)):
            body = body[1:-1]
        string = "(λ" + Abstraction.symbol_stack[0] + "." + body + ")"
        Abstraction.symbol_stack.pop(0)
        return string

    def beta(self, index, value):
        return Abstraction(self.expression.beta(index + 1, value))

    def shift_free(self, index, shift):
        return Abstraction(self.expression.shift_free(index + 1, shift))

class Application(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def reduce(self):
        if isinstance(self.left, Abstraction):
            body = self.left.expression
            body = body.beta(0, self.right.shift_free(0, 1))
            return body.shift_free(0, -1)

        reduced = self.left.reduce()
        if reduced is not None:
            return Application(reduced, self.right)

        reduced = self.right.reduce()
        if reduced is not None:
            return Application(self.left, reduced)

        return None

    def copy(self):
        return Application(self.left.copy(), self.right.copy())

    def __repr__(self):
        if isinstance(self.left, Variable) and isinstance(self.right, Variable):
            return "(" + repr(self.left) + " " + repr(self.right) + ")"
        elif isinstance(self.left, Application):
            if isinstance(self.right, Variable):
                return repr(self.left)[:-1] + " " + repr(self.right) + ")"
            else:
                return repr(self.left) + repr(self.right)
        return "(" + repr(self.left) + repr(self.right) + ")"

    def beta(self, index, value):
        left = self.left.beta(index, value)
        right = self.right.beta(index, value)
        return Application(left, right)

    def shift_free(self, index, shift):
        left = self.left.shift_free(index, shift)
        right = self.right.shift_free(index, shift)
        return Application(left, right)

class Variable(Expression):
    def __init__(self, index):
        self.index = index

    def reduce(self):
        return None

    def copy(self):
        return Variable(self.index)

    def __repr__(self):
        return Abstraction.symbol_stack[self.index]

    def beta(self, index, value):
        if self.index == index:
            return value.shift_free(0, index)
        return self.copy()

    def shift_free(self, index, shift):
        if self.index >= index:
            return Variable(self.index + shift)
        return self.copy()

def A(expr):
    if isinstance(expr, int):
        return Variable(expr)
    elif isinstance(expr, (list, tuple)):
        if len(expr) == 2:
            return Application(A(expr[0]), A(expr[1]))
        elif len(expr) == 1:
            return A(expr[0])
        return Application(A(expr[:-1]), A(expr[-1]))
    elif isinstance(expr, Expression):
        return expr

def L(expression):
    return Abstraction(A(expression))

def parse_string(string):
    if string.startswith("λ"):
        Abstraction.symbol_stack.insert(0, string[1])
        abstraction = L(parse_string(string[3:]))
        Abstraction.symbol_stack.pop(0)
        return abstraction
    elif len(string) == 1:
        return Abstraction.symbol_stack.index(string)

    applications = []
    while string:
        if string[0] == "(":
            end = 0
            brackets = 0
            for i, char in enumerate(string):
                if char == "(":
                    brackets += 1
                elif char == ")":
                    brackets -= 1
                if brackets == 0:
                    end = i
                    break
            applications.append(parse_string(string[1:end]))
            string = string[end + 1:]
        else:
            applications.append(Abstraction.symbol_stack.index(string[0]))
            string = string[1:]
            if string.startswith(" "):
                string = string[1:]
    return applications

def dump(expression, alternative=True, xlimit=None, ylimit=None):
    variables = []
    application_depths = []
    abstractions = {}
    abstraction_stack = []
    applications = []
    branching_points = []

    stack = [expression]
    while stack:
        expr = stack.pop()
        if isinstance(expr, str):
            if expr == "end_abstraction":
                start = abstraction_stack.pop()
                end = len(variables) - 1
                index = len(abstraction_stack)
                if index not in abstractions:
                    abstractions[index] = []
                abstractions[index].append((start, end))
            elif expr == "end_application":
                b = branching_points.pop()[0]
                if alternative:
                    a = branching_points.pop()[1]
                else:
                    a = branching_points.pop()[0]

                depth = max(
                    application_depths[a],
                    application_depths[b])
                applications.append([a, b, depth])
                application_depths[a] = depth + 1
                application_depths[b] = depth + 1
                branching_points.append((a, b))
        if isinstance(expr, Abstraction):
            abstraction_stack.append(len(variables))
            stack.append("end_abstraction")
            stack.append(expr.expression)
        elif isinstance(expr, Application):
            stack.append("end_application")
            stack.append(expr.right)
            stack.append(expr.left)
        elif isinstance(expr, Variable):
            variables.append(len(abstraction_stack) - expr.index - 1)
            application_depths.append(len(abstraction_stack))
            branching_points.append((len(variables) - 1, len(variables) - 1))

    if not alternative:
        application_depths[0] += 1

    # print("Variables:", variables)
    # print("Depths:", application_depths)
    # print("Abstractions:", abstractions)
    # print("Applications:", applications)

    width = len(variables) * 4
    height = max(application_depths) * 2
    canvas = [[0 for i in range(width)] for i in range(height)]

    for i in range(len(variables)):
        for j in range(variables[i] * 2, application_depths[i] * 2 - 1):
            canvas[j][i * 4 + 1] = 1

    for depth in abstractions:
        for abstraction in abstractions[depth]:
            start, end = abstraction
            canvas[depth * 2][start * 4:end * 4 + 3] = [1 for i in range((end - start) * 4 + 3)]

    for application in applications:
        start, end, depth = application
        y = depth * 2
        canvas[y][start * 4 + 1:end * 4 + 2] = [1 for i in range((end - start) * 4 + 1)]

    # pixels = [[" ", "▄"], ["▀", "█"]]
    # for i in range(0, height, 2):
    #     print("".join(pixels[canvas[i][x]][canvas[i+1][x]] for x in range(width)))

    pixels = [
        " 𜺨𜺫🮂𜴀▘𜴁𜴂𜴃𜴄▝𜴅𜴆𜴇𜴈▀", "𜴉𜴊𜴋𜴌🯦𜴍𜴎𜴏𜴐𜴑𜴒𜴓𜴔𜴕𜴖𜴗",
        "𜴘𜴙𜴚𜴛𜴜𜴝𜴞𜴟🯧𜴠𜴡𜴢𜴣𜴤𜴥𜴦", "𜴧𜴨𜴩𜴪𜴫𜴬𜴭𜴮𜴯𜴰𜴱𜴲𜴳𜴴𜴵🮅",
        "𜺣𜴶𜴷𜴸𜴹𜴺𜴻𜴼𜴽𜴾𜴿𜵀𜵁𜵂𜵃𜵄", "▖𜵅𜵆𜵇𜵈▌𜵉𜵊𜵋𜵌▞𜵍𜵎𜵏𜵐▛",
        "𜵑𜵒𜵓𜵔𜵕𜵖𜵗𜵘𜵙𜵚𜵛𜵜𜵝𜵞𜵟𜵠", "𜵡𜵢𜵣𜵤𜵥𜵦𜵧𜵨𜵩𜵪𜵫𜵬𜵭𜵮𜵯𜵰",
        "𜺠𜵱𜵲𜵳𜵴𜵵𜵶𜵷𜵸𜵹𜵺𜵻𜵼𜵽𜵾𜵿", "𜶀𜶁𜶂𜶃𜶄𜶅𜶆𜶇𜶈𜶉𜶊𜶋𜶌𜶍𜶎𜶏",
        "▗𜶐𜶑𜶒𜶓▚𜶔𜶕𜶖𜶗▐𜶘𜶙𜶚𜶛▜", "𜶜𜶝𜶞𜶟𜶠𜶡𜶢𜶣𜶤𜶥𜶦𜶧𜶨𜶩𜶪𜶫",
        "▂𜶬𜶭𜶮𜶯𜶰𜶱𜶲𜶳𜶴𜶵𜶶𜶷𜶸𜶹𜶺", "𜶻𜶼𜶽𜶾𜶿𜷀𜷁𜷂𜷃𜷄𜷅𜷆𜷇𜷈𜷉𜷊",
        "𜷋𜷌𜷍𜷎𜷏𜷐𜷑𜷒𜷓𜷔𜷕𜷖𜷗𜷘𜷙𜷚", "▄𜷛𜷜𜷝𜷞▙𜷟𜷠𜷡𜷢▟𜷣▆𜷤𜷥█",
    ]
    graph = ""
    for y in range(0, height, 4):
        if y > ylimit:
            break
        for x in range(0, width, 2):
            if x > xlimit:
                continue
            top = canvas[y][x] + canvas[y][x+1]*2 + canvas[y+1][x]*4 + canvas[y+1][x+1]*8
            if y == height - 2:  # height % 4 != 0
                bottom = 0
            else:
                bottom = canvas[y+2][x] + canvas[y+2][x+1]*2 + canvas[y+3][x]*4 + canvas[y+3][x+1]*8
            graph += pixels[bottom][top]
        graph += "\n"
    print(graph, end="")

s = L(L(L([[2, 0], [1, 0]])))
k = L(L(1))
i = L(0)

iota1 = A([s, [s, i, [k, s]], [k, k]])
s2 = A([iota1, [iota1, [iota1, [iota1, iota1]]]])
k2 = A([iota1, [iota1, [iota1, iota1]]])
i2 = A([iota1, iota1])

iota2 = A([s2, [s2, i2, [k2, s2]], [k2, k2]])

expr = parse_string("λx.((λy.y(y((λz.zz)(λz.λa.λb.(b(λc.λd.d))((λc.(((zz)c)((λd.dd)(λd.c(dd)))))(λc.λd.λe.λf.(fd)(e(ac)))))(λz.λa.λb.y(bz)))))(λy.λz.(z(λa.λb.a)y)))")
while expr:
    dump(expr, True, xlimit=250, ylimit=50)
    expr = expr.reduce()
